### ISSUE TITLE
add a builtin function for every llvm C library intrinsic and bit manipulation instrinsics

### ISSUE THREAD

**Issue by andrewrk** [zig-lang/zig#767]
http://llvm.org/docs/LangRef.html#standard-c-library-intrinsics
http://llvm.org/docs/LangRef.html#bit-manipulation-intrinsics

LLVM has these because
 * they can sometimes map to hardware instructions (such as sqrt) (see #681)
 * LLVM codegen assumes that it can call these functions, which is why we have to provide builtin.o

This issue is to
 * implement all the intrinsics in std/special/builtin.zig
 * add all the intrinsics as `@` builtin functions in zig
 * most std.math.* functions can just call out to the builtin functions

builtin.o is going to become #514

**Comment by jfo** [comment]
I like the idea of std library functions calling directly into intrinsics when possible, but that makes me wonder why the intrinsic would be available to user space at all, or at least why it would look the same as the other built-ins. 

The number of potential `@` functions is quite large, but there is no native name spacing scheme right now for keeping them grouped logically. Would that be something worth considering?

**Comment by andrewrk** [comment]
> I like the idea of std library functions calling directly into intrinsics when possible, but that makes me wonder why the intrinsic would be available to user space at all, or at least why it would look the same as the other built-ins.

The standard library is user space. The other built-ins are very similar. These are no different than `@ctz`, `@clz`, `@memset`, `@memcpy`, `@fence`, `@addWithOverflow`, etc.

> The number of potential @ functions is quite large, but there is no native name spacing scheme right now for keeping them grouped logically. Would that be something worth considering?

It's not that large. We're talking about adding 25 functions, and there are no name collisions, and we already use `@` to prevent collisions with user space functions.

**Comment by jfo** [comment]
Fair enough, but at the least it would help to group them in the documentation at some point, I think.

**Comment by emekoi** [comment]
# Standard C Library Intrinsics
- [x] [`llvm.memcpy`](https://ziglang.org/documentation/master/#memcpy)
- [ ] `llvm.memmove`
- [x] [`llvm.memset`](https://ziglang.org/documentation/master/#memset)
- [x] [`llvm.sqrt`](https://ziglang.org/documentation/master/#sqrt)
- [ ] `llvm.powi`
- [ ] `llvm.sin`
- [ ] `llvm.cos`
- [ ] `llvm.pow`
- [ ] `llvm.exp`
- [ ] `llvm.exp2`
- [ ] `llvm.log`
- [ ] `llvm.log10`
- [ ] `llvm.log2`
- [ ] `llvm.fma`
- [ ] `llvm.fabs`
- [ ] `llvm.minnum`
- [ ] `llvm.maxnum`
- [ ] `llvm.minimum`
- [ ] `llvm.maximum`
- [ ] `llvm.copysign`
- [ ] `llvm.floor`
- [ ] `llvm.ceil`
- [ ] `llvm.trunc`
## Already Implemented?
- [ ] [`llvm.rint`](https://ziglang.org/documentation/master/@floatToInt)
- [ ] [`llvm.nearbyint`](https://ziglang.org/documentation/master/@floatToInt)
- [ ] [`llvm.round`](https://ziglang.org/documentation/master/@floatToInt)
- [ ] [`llvm.lround`](https://ziglang.org/documentation/master/@floatToInt)
- [ ] [`llvm.llround`](https://ziglang.org/documentation/master/@floatToInt)

# Bit Manipulation Intrinsics
- [x] [`llvm.bitreverse`](https://ziglang.org/documentation/master/#bitReverse)
- [x] [`llvm.bswap`](https://ziglang.org/documentation/master/#byteSwap)
- [x] [`llvm.ctpop`](https://ziglang.org/documentation/master/#popCount)
- [x] [`llvm.ctlz`](https://ziglang.org/documentation/master/#clz)
- [x] [`llvm.cttz`](https://ziglang.org/documentation/master/#ctz)
- [ ] `llvm.fshl`
- [ ] `llvm.fshr`

**Comment by shawnl** [comment]
This requires adding a libm implementation to zig,(which i started here https://github.com/ziglang/zig/pull/2709) as llvm generates libcalls to libm. Also, once this is complete, I want to move implementations into `c.zig` and `libm.zig` and then have the std library versions just call the intrinsics, so that llvm (or other frontends) can optimize these as is appropriate.

**Comment by Tetralux** [comment]
Does LLVM also have a `pause` intrinsic?
https://www.felixcloutier.com/x86/pause

Would this be considered to be added for this too, or should that be a seperate issue?

**Comment by shawnl** [comment]
That one is a little tricky to do (if llvm already exports it which it
probably does) so I think you should open another issue.

**Comment by matu3ba** [comment]
semi-related to #1290, ie for powi

**Comment by achan1989** [comment]
std/special/builtin.zig no longer exists. Can I get a summary of where the intrinsics are implemented, and how they are exposed as `@` builtins?

**Comment by nektro** [comment]
mainly here
https://github.com/ziglang/zig/blob/master/lib/std/zig/BuiltinFn.zig
https://github.com/ziglang/zig/blob/master/lib/std/zig/Zir.zig#L133 (`std.zig.Zir.Inst.Tag`)
https://github.com/ziglang/zig/blob/master/src/Sema.zig#L1009 `fn analyzeBodyInner`
compile errors will guide you to everything else
